File: app/src/main/java/com/example/mg/masterdetail/UI/WeatherFragment.java
package com.example.mg.masterdetail.UI;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;
import com.example.mg.masterdetail.R;
import com.example.mg.masterdetail.UI.DI.DaggerIFragComponent;
import com.example.mg.masterdetail.UI.DI.FragmentModule;
import com.example.mg.masterdetail.WeatherApplication;
import com.example.mg.masterdetail.data.model.Weather10daysModel.ForecastBean.SimpleforecastBean.ForecastdayBeanX;
import com.example.mg.masterdetail.data.model.WeatherModel;
import com.example.mg.masterdetail.util.DialogUtils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;


public class WeatherFragment extends Fragment implements IDayListContract.IView {

    @Inject
    RequestManager glide;
    @Inject
    WeatherDayRecyclerViewAdapter adapter;
    @Inject
    DayListPresenter presenter;
    @Inject
    DialogUtils dialog;
    private ProgressDialog progressDialog;

    @BindView(R.id.text_city_name)
    TextView textCityName;
    @BindView(R.id.text_weather_condition)
    TextView textWeatherCondition;
    @BindView(R.id.image_weather_condition)
    ImageView imageWeatherCondition;
    @BindView(R.id.text_weather_temp)
    TextView textWeatherTemp;
    @BindView(R.id.toolbar_layout)
    CollapsingToolbarLayout toolbarLayout;
    @BindView(R.id.app_bar)
    AppBarLayout appBar;
    @BindView(R.id.text_conditon)
    TextView textCondition;
    @BindView(R.id.text_wind)
    TextView textWind;
    @BindView(R.id.text_humidity)
    TextView textHumidity;
    @BindView(R.id.linearLayout)
    LinearLayout linearLayout;
    @BindView(R.id.frameLayout)
    LinearLayout frameLayout;
    @BindView(R.id.text_no_internet)
    TextView textNoInternet;
    @BindView(R.id.day_list)
    RecyclerView recyclerView;

    Unbinder unbinder;
    //private static final String TAG = "WeatherFragment";
    private static final String KEY_TODAY = "KEY_TODAY";
    private static final String KEY_TEN_DAY = "KEY_TEN_DAY";
    private WeatherModel.CurrentObservationBean mTodayData;
    private List<ForecastdayBeanX> mDaysData;

    public WeatherFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DaggerIFragComponent.builder()
                .iAppComponent(WeatherApplication.get(getActivity()).getAppComponent())
                .fragmentModule(new FragmentModule(this))
                .build().inject(this);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_weather, container, false);
        unbinder = ButterKnife.bind(this, view);

        recyclerView.addItemDecoration(new DividerItemDecoration(
                recyclerView.getContext(),
                DividerItemDecoration.HORIZONTAL));

        if (savedInstanceState != null) {
            mTodayData = savedInstanceState.getParcelable(KEY_TODAY);
            mDaysData = savedInstanceState.getParcelableArrayList(KEY_TEN_DAY);
            init(mTodayData);
            setupRecyclerView(mDaysData);
        }
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.onResume();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(KEY_TODAY, mTodayData);
        outState.putParcelableArrayList(KEY_TEN_DAY, (ArrayList<? extends Parcelable>) mDaysData);

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        recyclerView.setAdapter(null);
        unbinder.unbind();
        presenter.onDestroy();
    }

    @Override
    public void setupRecyclerView(@NonNull List<ForecastdayBeanX> daysData) {
        mDaysData = daysData;
        recyclerView.setAdapter(adapter);
        adapter.setData(daysData);

    }

    @Override
    public void init(@NonNull WeatherModel.CurrentObservationBean todayData) {
        mTodayData = todayData;
        textCityName.setText(mTodayData.getDisplay_location().getFull());
        textWeatherCondition.setText(mTodayData.getIcon());
        textWeatherTemp.setText(mTodayData.getTemperature_string());

        textCondition.setText(String.format("condition\n%s", mTodayData.getIcon()));
        textWind.setText(String.format("wind speed\n%s kph", String.valueOf(mTodayData.getWind_kph())));
        textHumidity.setText(String.format("humidity\n%s", mTodayData.getRelative_humidity()));

        glide.asGif()
                .load(mTodayData.getIcon_url())
                .into(imageWeatherCondition);
    }

    @Override
    public void showNoInternet() {
        textNoInternet.setVisibility(View.VISIBLE);

    }

    @Override
    public void showLoading() {
        progressDialog = dialog.showLoadingDialog();
        progressDialog.show();

    }

    @Override
    public void hideLoading() {
        progressDialog.dismiss();
    }
}

File: app/src/main/java/com/example/mg/masterdetail/ui/WeatherFragment.kt
package com.example.mg.masterdetail.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import android.widget.Toast.LENGTH_SHORT
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import com.bumptech.glide.RequestManager
import com.example.mg.masterdetail.R
import com.example.mg.masterdetail.data.Result
import com.example.mg.masterdetail.data.model.WeatherModel
import kotlinx.android.synthetic.main.day_list.*
import kotlinx.android.synthetic.main.fragment_weather.*
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel
import java.util.*

class WeatherFragment : Fragment() {
    private val glide: RequestManager by inject()

    private val viewModel: DayListViewModel by viewModel()

    private val adapter: WeatherDayRecyclerViewAdapter by lazy {
        WeatherDayRecyclerViewAdapter(glide, ArrayList())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_weather, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        recyclerView.adapter = adapter
        recyclerView.addItemDecoration(DividerItemDecoration(
                recyclerView.context,
                DividerItemDecoration.HORIZONTAL))

        viewModel.weatherObserver.observe(viewLifecycleOwner, Observer {
            hideLoading()
            it?.data?.let { list ->
                init(list[0])
            }
        })

        viewModel.weatherDaysObserver.observe(viewLifecycleOwner, Observer {
            hideLoading()
            it?.data?.let { list ->
                adapter.setData(list)
            }
        })

        viewModel.errorObserver.observe(viewLifecycleOwner, Observer {
            hideLoading()
            when (it) {
                Result.Error.ServerError(null) -> Toast.makeText(requireContext(), "error in server", LENGTH_SHORT).show()
                Result.Error.InvalidUser -> TODO()
                Result.Error.NetworkError -> showNoInternet()
            }
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()
        recyclerView.adapter = null
    }


    @SuppressLint("SetTextI18n")
    private fun init(data: WeatherModel.Data) {
        textCityName.text = data.city_name
        textWeatherTemp.text = data.temp.toString()
        textWind.text = "wind speed\n${data.wind_spd} kph"
        textHumidity.text = "humidity\n${data.rh}"

        data.weather?.let {
            textCondition.text = "condition\n${it.description}"
            glide.load("https://www.weatherbit.io/static/img/icons/${it.icon}.png")
                    .into(imageWeatherCondition!!)
        }
    }

    private fun showNoInternet() {
        textNoInternet.visibility = View.VISIBLE
    }

    private fun hideLoading() {
        progressCircular.visibility = View.GONE
    }
}